### PR TITLE
Add detection library

### PR DESCRIPTION
**Pull request by dashpole**
This supersedes https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/383.

This adds a resource detection library, which provides a function for each resource attribute that can be detected by otel detectors.

### CODE REVIEW

File: detectors/gcp/gke.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+type LocationType int64
+
+const (
+	UndefinedLocation LocationType = iota
+	Zone
+	Region
+)
+
+// GKEAvaiabilityZoneOrRegion returns the location of the cluster and whether the cluster is zonal or regional
+func (d *Detector) GKEAvaiabilityZoneOrRegion() (string, LocationType, error) {
```
Comment (damemi) [inline detectors/gcp/gke.go:62]: typo: `Avaiability`

File: detectors/gcp/gce.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	return d.metadata.InstanceID()
+}
+
+// GCEHostName returns the instance name of the instance on which this program is running
+func (d *Detector) GCEHostName() (string, error) {
+	return d.metadata.InstanceName()
+}
+
+// GCEAvaiabilityZoneAndRegion returns the zone and region in which this program is running
+func (d *Detector) GCEAvaiabilityZoneAndRegion() (string, string, error) {
```
Comment (damemi) [inline detectors/gcp/gce.go:42]: typo: `Avaiability`

**Review by damemi — APPROVED** [review]

File: detectors/gcp/detector.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import (
+	"errors"
+	"os"
+
+	"cloud.google.com/go/compute/metadata"
+)
+
+var errEnvVarNotFound = errors.New("environment variable not found")
+
+// NewDetector returns a *Detector which can determine aspects of
```
[the remark this answers is not in the document]
Comment (dashpole) [reply detectors/gcp/detector.go]: fixec

File: detectors/gcp/faas.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	return "", errEnvVarNotFound
+}
+
+// FAASVersion returns the instance id of the cloud run instance or cloud function
+func (d *Detector) FAASInstanceID() (string, error) {
+	return d.metadata.InstanceID()
+}
+
+// FAASCloudRegion detects region from the metadata server.  It is in the
+// format /projects/123/regions/r.
```
[the remark this answers is not in the document]
Comment (dashpole) [reply detectors/gcp/faas.go]: updated

File: detectors/gcp/gke.go (hunk 1 above)
Comment (dashpole) [reply detectors/gcp/gke.go]: fixed

**Pull request merged**